enderio-base/src/main/java/com/enderio/base/common/capability/darksteel/DarkSteelUpgradeable.java

```diff
@@ -0,0 +1,173 @@
+package com.enderio.base.common.capability.darksteel;
+
+import com.enderio.base.EnderIO;
+import com.enderio.base.common.capability.EIOCapabilities;
+import com.enderio.base.common.item.darksteel.upgrades.DarkSteelUpgradeRegistry;
+import com.enderio.base.common.item.darksteel.upgrades.EmpoweredUpgrade;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.Tag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.ItemStack;
+
+import java.util.*;
+
+@SuppressWarnings("unused")
+public class DarkSteelUpgradeable implements IDarkSteelUpgradable {
+
+    //----------------------- Utils
+
+    public static ItemStack addUpgrade(ItemStack is, IDarkSteelUpgrade upgrade) {
+        is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).ifPresent(upgradable -> upgradable.addUpgrade(upgrade));
+        return is;
+    }
+
+    public static Collection<IDarkSteelUpgrade> getUpgrades(ItemStack is) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(IDarkSteelUpgradable::getUpgrades).orElse(Collections.emptyList());
+    }
+
+    public static boolean hasUpgrade(ItemStack is, String name) {
+        return is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).map(upgradable -> upgradable.hasUpgrade(name)).orElse(false);
+    }
+
+    public static <T extends IDarkSteelUpgrade> Optional<T> getUpgradeAs(ItemStack is, String upgrade, Class<T> as) {
+        Optional<IDarkSteelUpgradable> cap = is.getCapability(EIOCapabilities.DARK_STEEL_UPGRADABLE).resolve();
+        if(cap.isEmpty()) {
+            return Optional.empty();
+        }
+        return cap.get().getUpgradeAs(upgrade, as);
```

Comment:
```suggestion
        return cap.flatMap(upgradeCap -> upgradeCap.getUpgradeAs(upgrade, as));
```
use flatMap as it returns Optional.empty(), if cap is empty